### PR TITLE
PHOENIX-6453 Possible ArrayIndexOutOfBoundsException while preparing …

### PR DESCRIPTION
**Pull request by chrajeshbabu**
…scan start key with multiple key range queries(Rajeshbabu)

The field position preparation logic from the slot spans is wrong  so when there are mix of fixed length and variable length columns in different in lists scan key length coming wrong so getting AIOOB exception.

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  17m  5s |  master passed  |
| +0 |  hbaserecompile  |  24m  0s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  5s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 42s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 53s |  master passed  |
| +0 :ok: |  spotbugs  |   3m 21s |  phoenix-core in master has 959 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   9m 49s |  the patch passed  |
| +0 |  hbaserecompile  |  20m 13s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  3s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  3s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 42s |  phoenix-core: The patch generated 9 new + 698 unchanged - 5 fixed = 707 total (was 703)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 29s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 120m 44s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 17s |  The patch does not generate ASF License warnings.  |
|  |   | 180m 46s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1220 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux ff7ef1275cee 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 55aeb15 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/1/testReport/ |
| Max. process+thread count | 6706 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java [hunk 1]
```diff
@@ -363,10 +363,12 @@ public static void setTimeRange(Scan scan, long minStamp, long maxStamp) {
         }
         int[] position = new int[slots.size()];
         int maxLength = 0;
+        int slotEndingFieldPos = 0;
         for (int i = 0; i < position.length; i++) {
             position[i] = bound == Bound.LOWER ? 0 : slots.get(i).size()-1;
             KeyRange range = slots.get(i).get(position[i]);
-            Field field = schema.getField(i + slotSpan[i]);
+            slotEndingFieldPos = slotEndingFieldPos + slotSpan[i] + (i>0? 1: 0);
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java:370]: nit:

This is quite confusing : `+ (i>0? 1: 0);` .

Can we increment slotEndingFieldPos at the end of the for cycle body instead?
Or just initialize slotEndingFieldPos to -1 , and increment it here unconditionally ?

**Review by stoty — CHANGES_REQUESTED** [review]
The logic looks good, but I feel it could be made easier to read.

File: phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java (hunk 1 above)
Comment (joshelser) [reply phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java:370]: 👍🏼

**Review by stoty — APPROVED** [review]
+1 LGTM

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  7s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  15m 32s |  master passed  |
| +0 |  hbaserecompile  |  22m 21s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  6s |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 51s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 50s |  master passed  |
| +0 :ok: |  spotbugs  |   3m  3s |  phoenix-core in master has 965 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   8m  8s |  the patch passed  |
| +0 |  hbaserecompile  |  17m 33s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  3s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  3s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 50s |  phoenix-core: The patch generated 5 new + 698 unchanged - 5 fixed = 703 total (was 703)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 18s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 115m 19s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 47s |  The patch does not generate ASF License warnings.  |
|  |   | 169m  5s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.UpsertSelectIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1220 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux f94326a248ea 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 3231aad |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/2/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/2/testReport/ |
| Max. process+thread count | 11527 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1220/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by joshelser — APPROVED** [review]

**Pull request closed without merging**